### PR TITLE
Lint against needless uses of `collect()`

### LINKED ISSUE #3034: suggest iterator.count() instead of iterator.collect<Vec<_>>.len()
instead of doing something like
````rust
let len = iterator.collect<Vec<_>>.len();
````
we might want to suggest
````rust
let len = iterator.count();
````
however, ````.count()```` consumes the iterator so I am not sure if this is always applicable.

### PR DESCRIPTION
**Pull request by shssoichiro**
Handles cases of `.collect().len()`, `.collect().is_empty()`, and
`.collect().contains()`. This lint is intended to be generic enough to
be added to at a later time with other similar patterns that could be
optimized.

Closes #3034

### CODE REVIEW

File: clippy_lints/src/loops.rs [hunk 1]
```diff
@@ -2241,3 +2265,67 @@ impl<'a, 'tcx> Visitor<'tcx> for VarCollectorVisitor<'a, 'tcx> {
...
+        let iter = snippet(cx, args[0].span, "??");
+        return format!("{}.count()", iter);
+    }
+    unreachable!();
+}
+
+fn generate_needless_collect_is_empty_sugg<'a, 'tcx>(collect_expr: &'tcx Expr, cx: &LateContext<'a, 'tcx>) -> String {
+    if let ExprKind::MethodCall(_, _, ref args) = collect_expr.node {
+        let iter = snippet(cx, args[0].span, "??");
+        return format!("{}.any(|_| true)", iter);
```
Comment (oli-obk) [inline clippy_lints/src/loops.rs]: Is that really the best suggestion? Feels a little hacky

Comment (shssoichiro) [reply clippy_lints/src/loops.rs]: Hah, you're right. My mistake. I agree, `.next().is_none()` feels cleaner.

**Comment by shssoichiro** [comment]
> That's actually a false positive! It checks for the number of unique items in the iterator:

Looks like it. Ideally I think using `.unique().count()` from itertools is the better solution, but I'm not sure if clippy should be recommending external crates. Although it looks like internally, itertools uses a `HashMap` to track the seen items, so the performance difference is negligible at best in this case. I'll just update the lint to only look at `Vec` methods for now.

**Comment by killercup** [comment]
> I'll just update the lint to only look at Vec methods for now.

Sounds good for a start. Using other lists or maps should be fine, too, but sets have observably different properties.

**Comment by shssoichiro** [comment]
Rebased and fixed rustup issues.

**Pull request merged**